bot/cogs/token_remover.py

```diff
@@ -65,94 +74,105 @@ async def on_message_edit(self, before: Message, after: Message) -> None:
-        maybe_matches = TOKEN_RE.findall(msg.content)
-        for substr in maybe_matches:
-            if cls.is_maybe_token(substr):
+        for match in TOKEN_RE.finditer(msg.content):
+            token = Token(*match.groups())
+            if cls.is_valid_user_id(token.user_id) and cls.is_valid_timestamp(token.timestamp):
                 # Short-circuit on first match
-                return substr
+                return token
 
         # No matching substring
         return
 
-    @classmethod
-    def is_maybe_token(cls, test_str: str) -> bool:
-        """Check the provided string to see if it is a seemingly valid token."""
-        try:
-            user_id, creation_timestamp, hmac = test_str.split('.')
-        except ValueError:
-            return False
-
-        if cls.is_valid_user_id(user_id) and cls.is_valid_timestamp(creation_timestamp):
-            return True
-
     @staticmethod
     def is_valid_user_id(b64_content: str) -> bool:
         """
         Check potential token to see if it contains a valid Discord user ID.
 
         See: https://discordapp.com/developers/docs/reference#snowflakes
         """
-        b64_content += '=' * (-len(b64_content) % 4)
+        b64_content = utils.pad_base64(b64_content)
 
         try:
-            content: bytes = base64.b64decode(b64_content)
-            return content.decode('utf-8').isnumeric()
-        except (binascii.Error, UnicodeDecodeError):
+            decoded_bytes = base64.urlsafe_b64decode(b64_content)
+            string = decoded_bytes.decode('utf-8')
+
+            # isdigit on its own would match a lot of other Unicode characters, hence the isascii.
+            return string.isascii() and string.isdigit()
+        except (binascii.Error, ValueError):
             return False
 
     @staticmethod
     def is_valid_timestamp(b64_content: str) -> bool:
         """
-        Check potential token to see if it contains a valid timestamp.
+        Return True if `b64_content` decodes to a valid timestamp.
 
-        See: https://discordapp.com/developers/docs/reference#snowflakes
+        If the timestamp is greater than the Discord epoch, it's probably valid.
+        See: https://i.imgur.com/7WdehGn.png
         """
-        b64_content += '=' * (-len(b64_content) % 4)
+        b64_content = utils.pad_base64(b64_content)
 
         try:
-            content = base64.urlsafe_b64decode(b64_content)
-            snowflake = struct.unpack('i', content)[0]
-        except (binascii.Error, struct.error):
+            decoded_bytes = base64.urlsafe_b64decode(b64_content)
+            timestamp = int.from_bytes(decoded_bytes, byteorder="big")
+        except (binascii.Error, ValueError) as e:
+            log.debug(f"Failed to decode token timestamp '{b64_content}': {e}")
+            return False
+
+        # Seems like newer tokens don't need the epoch added, but add anyway since an upper bound
+        # is not checked.
```

Comment:
Do you have an example of a token like that?
Unsurprisingly I had no success finding any tokens online and all tokens I have available are direct timestamps but also quite new. Still, while not a big issue, this does potentially bring a few false positives like the `DISCORD_EPOCH - TOKEN_EPOCH` edge case which is invalid in the "new" format

Reply:
The image explaining the token format has an old token. I couldn't find any other old tokens. I think it's better to have false positives than false negatives, and this really is an edge case. On the other hand, old format tokens are going to be infrequent so we could just stop checking for them.